Make HeapFloat32Array generic

Taym95 opened the pull request:
part #31064

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because there is not behavior change

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Taym95 commented on components/script/dom/bindings/typedarrays.rs:
```diff
@@ -9,30 +9,36 @@ use std::ptr;
 
 #[derive(Default, JSTraceable)]
-pub struct HeapFloat32Array {
+pub struct HeapTypedArray {
```
@gterzian please review, I tried to make it HeapTypedArray<T> but since T is not used here I added `PhantomData ` but I got errors because of `JSTraceable `

sagudev replied:
on phantom field you need to add `#[no_trace]`

Taym95 committed:
Make HeapFloat32Array generic

Signed-off-by: Bentaimia Haddadi <haddadi.taym@gmail.com>

gterzian reviewed (changes requested):
Great work, thanks! I think it's a very good idea to add a phantom type, and make it no trace as described at https://github.com/servo/servo/pull/31167/files#r1465417408

Can this phantom type also be used as the generic parameter to the various methods? 

Perhaps an alternative would be a trait `HeapTypedArray`, with a `type ArrayType;`, and then have specific structs implement the trait but only by setting the `ArrayType`(with all methods being defaults). So for example `HeapFloat32Array` would be left as is, but only implement `HeapTypedArray` and set the type to `Float32`, or something along those lines.

Taym95 commented:
> Great work, thanks! I think it's a very good idea to add a phantom type, and make it no trace as described at https://github.com/servo/servo/pull/31167/files#r1465417408
> 
> Can this phantom type also be used as the generic parameter to the various methods?
> 
> Perhaps an alternative would be a trait `HeapTypedArray`, with a `type ArrayType;`, and then have specific structs implement the trait but only by setting the `ArrayType`(with all methods being defaults). So for example `HeapFloat32Array` would be left as is, but only implement `HeapTypedArray` and set the type to `Float32`, or something along those lines.

I added `HeapTypedArray ` trait and implemented the trait for `HeapFloat32Array`

sagudev commented:
I think using traits will require code duplication for each type, so if possible I would peruse generic approach using phantom type and as far as I know same generic can also be required in methods.

gterzian commented:
> I added HeapTypedArray trait and implemented the trait for HeapFloat32Array

Thank you!

Is it possible to make all the methods defaults on the trait, using `Self::ArrayType`, so that the actual implementation of the trait is only used to set `ArrayType` and is otherwise empty(If the `typedarray` macro doesn't like it, I would drop the macro)? If not, I agree with @sagudev that using a phantom type is better.

Taym95 commented:
> I think using traits will require code duplication for each type, so if possible I would peruse generic approach using phantom type and as far as I know same generic can also be required in methods.

I added all the methods defaults on the trait to avoid  code duplication for each type, could not use `PhantomData<T>` with `#[no_trace]` because of:

`cannot find type `T` in this scope
not found in this scope`

sagudev commented:
> could not use `PhantomData<T>` with `#[no_trace]` because of:
> 
> `cannot find type T in this scope not found in this scope`

Do you have WIP branch of it, I would like to see what macro expansion does.

Manually implementing JSTraceable trait should work.

Taym95 commented:
> > could not use `PhantomData<T>` with `#[no_trace]` because of:
> > `cannot find type T in this scope not found in this scope`
> 
> Do you have WIP branch of it, I would like to see what macro expansion does.
> 
> Manually implementing JSTraceable trait should work.

somehow macro not handling generic type, branch:

https://github.com/Taym95/servo/tree/make-HeapFloat32Array-generic-phantom

Taym95 committed:
Define all the methods defaults on the HeapTypedArray trait

Signed-off-by: Bentaimia Haddadi <haddadi.taym@gmail.com>

sagudev commented:
> somehow macro not handling generic type

Yes, indeed.

For such cases manual implementation is needed:
```rust
#[allow(unsafe_code)]
unsafe impl<T> crate::dom::bindings::trace::JSTraceable for HeapTypedArray<T> {
    #[inline]
    unsafe fn trace(&self, tracer: *mut js::jsapi::JSTracer) {
        self.internal.trace(tracer);
    }
}
```

sagudev commented:
Also `TypedArrayElementCreator + TypedArrayElement` constraints on generic might be more sensible directly on struct IMO.

Taym95 committed:
Use generic type instead of trait

Signed-off-by: Bentaimia Haddadi <haddadi.taym@gmail.com>

Taym95 commented:
> > somehow macro not handling generic type
> 
> Yes, indeed.
> 
> For such cases manual implementation is needed:
> 
> ```rust
> #[allow(unsafe_code)]
> unsafe impl<T> crate::dom::bindings::trace::JSTraceable for HeapTypedArray<T> {
>     #[inline]
>     unsafe fn trace(&self, tracer: *mut js::jsapi::JSTracer) {
>         self.internal.trace(tracer);
>     }
> }
> ```

Thanks @sagudev I used generic type and last commit. and kept old trait commit

gterzian commented on components/script/dom/bindings/typedarrays.rs:
```diff
@@ -4,40 +4,59 @@
 
-impl HeapFloat32Array {
-    pub fn set_data(&self, cx: JSContext, data: &[f32]) -> Result<(), ()> {
+#[allow(unsafe_code)]
```
This is unnecessary because the entire module allows unsafe code(see top of file).

gterzian commented on components/script/dom/bindings/typedarrays.rs:
```diff
@@ -4,40 +4,59 @@
             unsafe {
                 rooted!(in (*cx) let view_buffer =
-                    JS_GetArrayBufferViewBuffer(*cx, self.internal.handle(), &mut is_shared));
+                     JS_GetArrayBufferViewBuffer(*cx, self.internal.handle(), &mut is_shared));
```
is this an fmt change?

Taym95 replied:
no I will fix it

gterzian commented on components/script/dom/bindings/typedarrays.rs:
```diff
@@ -103,16 +127,15 @@ impl HeapFloat32Array {
 where
-    T: TypedArrayElementCreator + TypedArrayElement,
-    S: JSObjectStorage,
+    T: TypedArrayElement + TypedArrayElementCreator,
```
Why no leave the generic `JSObjectStorage` bound and use `S` instead of `JSObject`?

Taym95 replied:
to avoid adding generic `S` to create_typed_array when I call it from `set_data` or `TextEncoder`

gterzian reviewed (changes requested):
great work, thanks! Just one nit and some questions.

gterzian reviewed (approved):
LGTM, thanks! And thanks for your help @sagudev
